funsor/tensor.py

```diff
@@ -772,6 +772,25 @@ def eager_reshape_tensor(op, arg):
     return Tensor(data, arg.inputs, arg.dtype)
 
 
+@eager.register(Unary, ops.SumOp, Tensor)
+def eager_reshape_tensor(op, arg):
+    if not arg.inputs:
+        return Tensor(op(arg.data), arg.inputs, arg.dtype)
+
+    # Work around batch inputs.
```

Comment:
This logic converting dimensions into batch-aware dimensions seems useful and general-purpose enough that some version of it should maybe live in an `Op` method or something - we don't want to have to write this from scratch in each new op with nontrivial shape semantics.

Reply:
I agree this is useful, and I'm hoping @ordabayevy will generalize this in #482 once we have more than one use case.  This file seems like the right place for that general logic since `funsor.ops` should be agnostic to `inputs` and `domains` etc.